File: adapters/go/inspector/inspector.go
// Package inspector provides operator-visibility helpers over the explicit
// query and point-store interfaces.
//
// Cross-run listing requires a QueryStore. Reset helpers and run-scoped lists
// use the authoritative point Store.
package inspector

import (
	"context"
	"fmt"

	temporalessv1 "github.com/jim-technologies/temporaless/core/go/gen/temporaless/v1"
	"github.com/jim-technologies/temporaless/core/go/storage"
)

// ListInFlightWorkflows returns every workflow record whose status is
// WORKFLOW_STATUS_IN_PROGRESS.
func ListInFlightWorkflows(ctx context.Context, query storage.WorkflowQueryStore) ([]*temporalessv1.WorkflowRecord, error) {
	return listWorkflowsByStatus(ctx, query, temporalessv1.WorkflowStatus_WORKFLOW_STATUS_IN_PROGRESS)
}

// ListFailedWorkflows returns every workflow record whose status is
// WORKFLOW_STATUS_FAILED.
func ListFailedWorkflows(ctx context.Context, query storage.WorkflowQueryStore) ([]*temporalessv1.WorkflowRecord, error) {
	return listWorkflowsByStatus(ctx, query, temporalessv1.WorkflowStatus_WORKFLOW_STATUS_FAILED)
}

// ListWorkflowsByStatus is the generic form, exposed for callers that want a
// status the helpers above don't cover (e.g. COMPLETED for audits).
func ListWorkflowsByStatus(
	ctx context.Context,
	query storage.WorkflowQueryStore,
	status temporalessv1.WorkflowStatus,
) ([]*temporalessv1.WorkflowRecord, error) {
	return listWorkflowsByStatus(ctx, query, status)
}

func listWorkflowsByStatus(ctx context.Context, query storage.WorkflowQueryStore, status temporalessv1.WorkflowStatus) ([]*temporalessv1.WorkflowRecord, error) {
	if query == nil {
		return nil, fmt.Errorf("query store is required")
	}
	var records []*temporalessv1.WorkflowRecord
	pageToken := ""
	seenPageTokens := map[string]struct{}{}
	for {
		response, err := query.ListWorkflows(ctx, &temporalessv1.ListWorkflowsRequest{
			Status:    status,
			PageToken: pageToken,
		})
		if err != nil {
			return nil, err
		}
		records = append(records, response.GetRecords()...)
		next := response.GetNextPageToken()
		if next == "" {
			return records, nil
		}
		if _, repeated := seenPageTokens[next]; repeated {
			return nil, fmt.Errorf("query store repeated page token %q", next)
		}
		seenPageTokens[next] = struct{}{}
		pageToken = next
	}
}

// ListActivities returns every activity record under the given workflow run.
func ListActivities(
	ctx context.Context,
	store storage.Store,
	key storage.WorkflowKey,
) ([]*temporalessv1.ActivityRecord, error) {
	if store == nil {
		return nil, fmt.Errorf("store is required")
	}
	return store.ListActivities(ctx, key)
}

// ResetWorkflow deletes the workflow record so the next invocation re-executes
// from scratch. Activity, timer, event, and claim records under the same run
// are left untouched — call ResetActivity or use a new run_id if a full reset
// is intended.
func ResetWorkflow(ctx context.Context, store storage.Store, key storage.WorkflowKey) error {
	if store == nil {
		return fmt.Errorf("store is required")
	}
	_, err := store.DeleteWorkflow(ctx, key)
	return err
}

// ResetActivity deletes a stored activity record so the next ExecuteActivity
// call re-executes the activity body.
func ResetActivity(ctx context.Context, store storage.Store, key storage.ActivityKey) error {
	if store == nil {
		return fmt.Errorf("store is required")
	}
	_, err := store.DeleteActivity(ctx, key)
	return err
}

// ResetEvent deletes a stored event record so the workflow's WaitEvent call
// returns ErrEventPending again on the next invocation.
func ResetEvent(ctx context.Context, store storage.Store, key storage.EventKey) error {
	if store == nil {
		return fmt.Errorf("store is required")
	}
	_, err := store.DeleteEvent(ctx, key)
	return err
}

File: core/py/src/temporaless/inspector.py
"""Operator-visibility helpers.

Cross-run listing is served by an optional query index. Reset helpers remain
point deletes against the core Store.
"""

from __future__ import annotations

from temporaless.storage import ActivityKey, EventKey, QueryStore, Store, WorkflowKey
from temporaless.v1 import temporaless_pb2


async def list_in_flight_workflows(query: QueryStore) -> list[temporaless_pb2.WorkflowRecord]:
    """Return every workflow record whose status is IN_PROGRESS."""
    return await _list_all_workflows(query, temporaless_pb2.WORKFLOW_STATUS_IN_PROGRESS)


async def list_failed_workflows(query: QueryStore) -> list[temporaless_pb2.WorkflowRecord]:
    """Return every workflow record whose status is FAILED."""
    return await _list_all_workflows(query, temporaless_pb2.WORKFLOW_STATUS_FAILED)


async def list_workflows_by_status(
    query: QueryStore,
    status: temporaless_pb2.WorkflowStatus,
) -> list[temporaless_pb2.WorkflowRecord]:
    """Generic form, exposed for callers wanting COMPLETED audits and similar."""
    return await _list_all_workflows(query, status)


async def _list_all_workflows(
    query: QueryStore,
    status: temporaless_pb2.WorkflowStatus,
) -> list[temporaless_pb2.WorkflowRecord]:
    records: list[temporaless_pb2.WorkflowRecord] = []
    page_token = ""
    seen_tokens: set[str] = set()
    while True:
        page, next_page_token = await query.list_workflows("", "", status, page_token=page_token)
        records.extend(page)
        if not next_page_token:
            return records
        if next_page_token == page_token or next_page_token in seen_tokens:
            raise RuntimeError("workflow query returned a repeated page token")
        seen_tokens.add(next_page_token)
        page_token = next_page_token


async def list_activities(store: Store, key: WorkflowKey) -> list[temporaless_pb2.ActivityRecord]:
    """Return every activity record under the given workflow run."""
    return await store.list_activities(key)


async def reset_workflow(store: Store, key: WorkflowKey) -> None:
    """Delete the workflow record so the next invocation re-executes from scratch."""
    await store.delete_workflow(key)


async def reset_activity(store: Store, key: ActivityKey) -> None:
    """Delete the activity record so the next ExecuteActivity re-executes."""
    await store.delete_activity(key)


async def reset_event(store: Store, key: EventKey) -> None:
    """Delete the event record so WaitEvent returns pending again."""
    await store.delete_event(key)
